--- /home/rossierd/soo/micofe/build/tmp/work/usr-linux-1.0-r0/usr-linux-1.0/src/soo/emiso_engine/webserver/plugins/plugins.cpp	1970-01-01 01:00:00.000000000 +0100
+++ ./src/soo/emiso_engine/webserver/plugins/plugins.cpp	2025-10-10 17:28:33.594220426 +0200
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 2023 Jean-Pierre Miceli <jean-pierre.miceli@heig-vd.ch>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#include "plugins.hpp"
+
+namespace emiso {
+namespace plugins {
+
+    PluginsApi::PluginsApi(httpserver::webserver *server): _server(server) {
+        std::string path  = "/plugins";
+
+        // Create routes and handlers
+        _listHandler = new ListHandler();
+
+        _server->register_resource(path, _listHandler);
+        _server->register_resource("/v[1-9]+.[0-9]+" + path, _listHandler);
+    }
+
+    PluginsApi::~PluginsApi() {}
+
+} // plugins
+} // emiso
