--- /home/rossierd/edgemtech/products/edgem1/build/tmp/work/uboot-2024.07-r0/uboot-2024.07/arch/arm/mach-imx/imx8m/Kconfig	2025-11-21 14:56:51.328620960 +0100
+++ ./arch/arm/mach-imx/imx8m/Kconfig	2025-11-21 14:54:55.153969190 +0100
@@ -257,6 +257,12 @@
 	select IMX8MQ
 	select IMX8M_LPDDR4
 
+config TARGET_TORADEX_SMARC_IMX8MP
+	bool "Support Toradex SMARC iMX8M Plus module"
+	select IMX8MP
+	select SUPPORT_SPL
+	select IMX8M_LPDDR4
+
 config TARGET_VERDIN_IMX8MM
 	bool "Support Toradex Verdin iMX8M Mini module"
 	select IMX8MM
@@ -373,6 +379,7 @@
 source "board/ronetix/imx8mq-cm/Kconfig"
 source "board/technexion/pico-imx8mq/Kconfig"
 source "board/variscite/imx8mn_var_som/Kconfig"
+source "board/toradex/smarc-imx8mp/Kconfig"
 source "board/toradex/verdin-imx8mm/Kconfig"
 source "board/toradex/verdin-imx8mp/Kconfig"
 
