--- /home/rossierd/soo/git/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/arch/arm64/boot/dts/arm/virt64_guest.dts	2025-08-30 15:42:55.089796760 +0200
+++ ./arch/arm64/boot/dts/arm/virt64_guest.dts	2025-08-29 08:46:39.933578397 +0200
@@ -351,13 +351,15 @@
 			compatible = "arm,cortex-a72";
 			device_type = "cpu";
 		};
+#if 0
 
 		cpu@3 {
 			reg = <0x03>;
 			enable-method = "psci";
 			compatible = "arm,cortex-a72";
 			device_type = "cpu";
-		};	
+		};
+#endif	
 	};
 
 	timer {
@@ -421,5 +423,37 @@
 			};
 		};
 	};
+
+
+	agency: agency {
+				
+		backends {
+
+			vdummy {
+				compatible = "vdummy,backend";
+				status = "disabled";
+			};
+	
+			vuart {
+				compatible = "vuart,backend";
+				status = "ok";
+			};
+			
+			vsenseled {
+            			compatible = "vsenseled,backend";
+            			status = "disabled";
+        		};
+			
+        		vsensej {
+            			compatible = "vsensej,backend";
+            			status = "disabled";
+        		};
+			
+			vlogs {
+				compatible = "vlogs,backend";
+				status = "ok";
+			};
+		};
+	};
 
 };
