--- /home/rossierd/soo/git/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/soo/lib/ctype.c	1970-01-01 01:00:00.000000000 +0100
+++ ./soo/lib/ctype.c	2025-08-29 08:46:41.395518498 +0200
@@ -0,0 +1,65 @@
+
+#include <linux/ctype.h>
+
+#include <soo/ctype.h>
+
+int __isalpha(int c)
+{
+	return isalpha(c);
+}
+
+int __isdigit(int c)
+{
+	return isdigit(c);
+}
+
+int __isalnum(int c)
+{
+	return isalnum(c);
+}
+
+int isblank(int c)
+{
+	return (c == ' ' || c == '\t');
+}
+
+int __iscntrl(int c)
+{
+	return iscntrl(c);
+}
+
+int __isgraph(int c)
+{
+	return isgraph(c);
+}
+
+int __islower(int c)
+{
+	return islower(c);
+}
+
+int __isprint(int c)
+{
+	return isprint(c);
+}
+
+int __ispunct(int c)
+{
+	return ispunct(c);
+}
+
+int __isspace(int c)
+{
+	return isspace(c);
+}
+
+int __isupper(int c)
+{
+	return isupper(c);
+}
+
+int __isxdigit(int c)
+{
+	return isxdigit(c);
+}
+
