--- /home/rossierd/soo/infrabase/build/tmp/work/usr-linux-1.0-r0/usr-linux-1.0/src/graphic/fb_benchmark/CMakeLists.txt	1970-01-01 01:00:00.000000000 +0100
+++ ./src/graphic/fb_benchmark/CMakeLists.txt	2025-08-28 13:15:04.926390816 +0200
@@ -0,0 +1,31 @@
+project(fb_benchmark)
+set(CMAKE_C_FLAGS "-g")
+
+add_executable(fb_benchmark
+    main.c
+    common.c
+    drm-atomic.c
+    drm-common.c
+    drm-legacy.c
+    draw.c
+    fb.c
+)
+
+# Un-Comment this block if building with IB.
+# TODO detect if building from IB.
+
+#find_package(PkgConfig REQUIRED)
+#target_link_libraries(fb_benchmark -Wl,-lgbm drm)
+
+# Un-Comment this block if building with the Docker container
+# It uses pkg-config to find the arguments to pass to the linker and includes paths
+
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(LIBDRM REQUIRED libdrm)
+pkg_check_modules(LIBGBM REQUIRED gbm)
+
+target_include_directories(fb_benchmark PUBLIC ${LIBDRM_INCLUDE_DIRS} ${LIBGBM_INCLUDE_DIRS})
+
+target_link_libraries(fb_benchmark ${LIBDRM_LIBRARIES} ${LIBGBM_LIBRARIES})
+
+
