--- a/package/Config.in
+++ b/package/Config.in
@@ -620,6 +620,10 @@
 	source "package/sdparm/Config.in"
 	source "package/sedutil/Config.in"
 	source "package/setserial/Config.in"
+	if !BR2_PACKAGE_BUSYBOX_HIDE_OTHERS
+	# busybox has fdisk
+	source "package/sfdisk/Config.in"
+	endif
 	source "package/sg3_utils/Config.in"
 	source "package/sigrok-cli/Config.in"
 	source "package/sispmctl/Config.in"
@@ -2756,6 +2760,9 @@
 	source "package/apg/Config.in"
 	source "package/at/Config.in"
 	source "package/bash-completion/Config.in"
+
+	source "package/auto-getty/Config.in"
+
 	source "package/catatonit/Config.in"
 	source "package/ccrypt/Config.in"
 	source "package/crudini/Config.in"
