--- /home/rossierd/edgemtech/customer/cybelec_linux/cybelec_linux/ib/build/tmp/work/usr-linux-1.0-r0/usr-linux-1.0/src/evl.c	1970-01-01 01:00:00.000000000 +0100
+++ ./src/evl.c	2025-12-08 08:25:37.313961945 +0100
@@ -0,0 +1,48 @@
+/**
+ * EVL user app
+ */
+
+#define _GNU_SOURCE
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <unistd.h>
+#include <evl/clock.h>
+#include <pthread.h>
+#include <evl/fcntl.h> /* Needs to be before #include <fcntl.h> */
+#include <evl/thread.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <errno.h>
+
+#include <evl/evl.h>
+#include <error.h>
+#include <sched.h>
+
+#define EVL_RT_CPU 1
+
+int main(int argc, char *argv[]) {
+	int ret, tfd;
+	cpu_set_t cpuset;
+	struct sched_param param;
+
+	evl_init();
+
+        param.sched_priority = 8;
+        ret = pthread_setschedparam(pthread_self(), SCHED_FIFO, &param);
+        if (ret) {
+                perror("Pthread setschedparam() failed");
+                return EXIT_FAILURE;
+        }
+          
+        CPU_ZERO(&cpuset);
+        CPU_SET(EVL_RT_CPU, &cpuset);
+        sched_setaffinity(getpid(), sizeof(cpuset), &cpuset);
+
+        tfd = evl_attach_self("evl_thread-%d", getpid());
+        if (tfd < 0) {
+                perror("EVL attatch self failed\n");
+		return EXIT_FAILURE;
+	}
+}
\ No newline at end of file
