--- /home/rossierd/edgemtech/customer/cybelec_linux/cybelec_linux/ib/build/tmp/work/usr-linux-1.0-r0/usr-linux-1.0/src/CMakeLists.txt	2025-12-09 17:53:46.420339216 +0100
+++ ./src/CMakeLists.txt	2025-12-09 17:56:56.586354130 +0100
@@ -1,6 +1,15 @@
 


 
+include_directories(
+    ../lib/evl/include
+	../lib/evl/build/include
+)
 
+add_executable(evl evl.c)
+
+target_link_libraries(evl 
+    PRIVATE
+        ${CMAKE_SOURCE_DIR}/lib/evl/build/lib/libevl.a)
 
 
