--- /home/rossierd/soo/git/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/include/linux/tty_flip.h	2024-11-17 23:15:08.000000000 +0100
+++ ./include/linux/tty_flip.h	2025-08-29 08:46:41.279848444 +0200
@@ -5,6 +5,10 @@
 #include <linux/tty_buffer.h>
 #include <linux/tty_port.h>
 
+#ifdef CONFIG_SOO
+#include <soo/avzcons.h>
+#endif /* CONFIG_SOO */
+
 struct tty_ldisc;
 
 int tty_buffer_set_limit(struct tty_port *port, int limit);
@@ -67,6 +71,12 @@
 	struct tty_buffer *tb = port->buf.tail;
 	int change;
 
+#ifdef CONFIG_SOO
+	/* Only chars issued from the main console are considered */
+	if (port->console && avz_switch_console(ch))
+ 		return 1;
+#endif
+
 	change = !tb->flags && (flag != TTY_NORMAL);
 	if (!change && tb->used < tb->size) {
 		if (tb->flags)
