--- /home/rossierd/soo/git/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/init/main.c	2024-11-17 23:15:08.000000000 +0100
+++ ./init/main.c	2025-08-29 08:46:41.312755195 +0200
@@ -109,6 +109,10 @@
 #include <asm/sections.h>
 #include <asm/cacheflush.h>
 
+#ifdef CONFIG_SOO
+#include <soo/hypervisor.h>
+#endif /* CONFIG_SOO */
+
 #define CREATE_TRACE_POINTS
 #include <trace/events/initcall.h>
 
@@ -994,6 +998,13 @@
 	 */
 	workqueue_init_early();
 
+#ifdef CONFIG_SOO
+
+	/* Get the reference to the AVZ shared data */
+	avz_get_shared();
+
+#endif /* CONFIG_SOO */
+
 	rcu_init();
 
 	/* Trace events are available after this */
