--- /home/rossierd/soo/git/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/Makefile	2025-08-30 15:42:55.070030785 +0200
+++ ./Makefile	2025-08-29 08:46:41.335442977 +0200
@@ -738,6 +738,10 @@
 # Defaults to vmlinux, but the arch makefile usually adds further targets
 all: vmlinux
 
+ifdef CONFIG_SOO
+LINUXINCLUDE += -Isoo/include
+endif
+
 CFLAGS_GCOV	:= -fprofile-arcs -ftest-coverage
 ifdef CONFIG_CC_IS_GCC
 CFLAGS_GCOV	+= -fno-tree-loop-im
@@ -1104,6 +1108,10 @@
 export MODORDER := $(extmod_prefix)modules.order
 export MODULES_NSDEPS := $(extmod_prefix)modules.nsdeps
 
+ifdef CONFIG_SOO
+  core-y	 += soo/
+endif
+
 ifeq ($(KBUILD_EXTMOD),)
 
 build-dir	:= .
