--- /home/rossierd/soo/git/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/soo/Kconfig	1970-01-01 01:00:00.000000000 +0100
+++ ./soo/Kconfig	2025-08-29 08:46:41.396316962 +0200
@@ -0,0 +1,16 @@
+
+config SOO
+	bool "Enabling the Smart Object Oriented (SOO) subsystem"
+
+if SOO
+menu "SOO Framework and subsystems"
+
+config SOO_CORE_ASF
+	bool "Enabling the security framework capabilities"
+	depends on ARM_PSCI
+
+source "soo/drivers/Kconfig"
+
+endmenu # SOO
+
+endif
