--- /home/rossierd/soo/git/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/soo/lib/libroxml/Makefile	1970-01-01 01:00:00.000000000 +0100
+++ ./soo/lib/libroxml/Makefile	2025-08-29 08:46:41.395330459 +0200
@@ -0,0 +1,12 @@
+obj-y += roxml_buff.o roxml_commit.o roxml_content.o 
+obj-y += roxml_core.o roxml_edit.o roxml_mem.o roxml_nav.o
+obj-y += roxml_parser.o roxml_xpath.o roxml_file.o
+
+ifeq ($(ARCH),arm)
+
+CFLAGS_roxml_xpath.o += -march=armv7-a -mfpu=neon -mfloat-abi=softfp
+else ifeq ($(ARCH),arm64)
+
+CFLAGS_REMOVE_roxml_xpath.o = -mgeneral-regs-only
+
+endif
