diff -Naru ../filesystem/create_img.sh ./create_img.sh
--- ../filesystem/create_img.sh	2022-12-21 17:53:24.290187815 +0100
+++ ./create_img.sh	2023-03-09 12:21:10.541062492 +0100
@@ -14,7 +14,7 @@
     echo Creating sdcard.img.$1 ... 
     dd_size=256M
     dd if=/dev/zero of=sdcard.img.$1 bs="$dd_size" count=1
-    devname=$(sudo losetup --partscan --find --show sdcard.img.$1)
+    devname=$(losetup --partscan --find --show sdcard.img.$1)
 
     # Keep device name only without /dev/
     devname=${devname#"/dev/"}
@@ -27,7 +27,7 @@
 
 if [ "$1" == "vexpress" -o "$1" == "rpi4" -o "$1" == "rpi4_64" -o "$1" == "virt64" ]; then
 #create the partition layout this way
-    (echo o; echo n; echo p; echo; echo; echo +128M; echo t; echo c; echo n; echo p; echo; echo; echo +400M; echo n; echo p; echo; echo; echo +100M; echo n; echo p; echo; echo; echo; echo w)   | sudo fdisk /dev/"$devname";
+    (echo o; echo n; echo p; echo; echo; echo +128M; echo t; echo c; echo n; echo p; echo; echo; echo +400M; echo n; echo p; echo; echo; echo +100M; echo n; echo p; echo; echo; echo; echo w)   | fdisk /dev/"$devname";
 fi
 
 echo Waiting...
@@ -38,10 +38,10 @@
     export devname="${devname}p"
 fi
 
-sudo mkfs.fat -F32 -v /dev/"$devname"1
-sudo mkfs.ext4 /dev/"$devname"2
+mkfs.fat -F32 -v /dev/"$devname"1
+mkfs.ext4 /dev/"$devname"2
 
 if [ "$1" == "vexpress" -o "$1" == "virt64" ]; then
-	sudo losetup -D
+	losetup -D
 fi
 
diff -Naru ../filesystem/mount.sh ./mount.sh
--- ../filesystem/mount.sh	2022-12-21 17:53:24.290187815 +0100
+++ ./mount.sh	2023-03-09 12:21:22.692774241 +0100
@@ -10,18 +10,18 @@
     PLATFORM=$2
 fi
 
-sudo rm -rf fs/*
+rm -rf fs/*
 mkdir -p fs
  
 if [ "$PLATFORM" == "vexpress" -o "$PLATFORM" == "virt64" ]; then
-    devname=$(sudo losetup --partscan --find --show sdcard.img.${PLATFORM})
+    devname=$(losetup --partscan --find --show sdcard.img.${PLATFORM})
     FS_IMG=sdcard.img.${PLATFORM}
 
-    #sudo losetup -P --find --show flash
+    #losetup -P --find --show flash
    
 
     # device is loopback (loop<n>)
-    sudo mount ${devname}p$1 fs 
+    mount ${devname}p$1 fs 
     exit 0
 fi
 
@@ -35,5 +35,5 @@
 fi
 
 if [ "$PLATFORM" == "rpi4" -o "$PLATFORM" == "rpi4_64" ]; then
-    sudo mount /dev/"${devname}"$1 fs
+    mount /dev/"${devname}"$1 fs
 fi
diff -Naru ../filesystem/umount.sh ./umount.sh
--- ../filesystem/umount.sh	2022-12-21 17:53:24.290187815 +0100
+++ ./umount.sh	2023-03-09 12:21:29.460613718 +0100
@@ -12,11 +12,11 @@
 
 sleep 1
 
-sudo umount fs
+umount fs
 
 # Let the filesystem be synchronized
 sleep 1
 
 if [ "$PLATFORM" == "vexpress" -o "$PLATFORM" == "virt64" ]; then
-    sudo losetup -D
+    losetup -D
 fi
