diff -Naru ../rootfs/deploy.sh ./deploy.sh
--- ../rootfs/deploy.sh	2022-12-22 16:32:13.517527284 +0100
+++ ./deploy.sh	2023-03-09 12:15:09.089667137 +0100
@@ -15,8 +15,8 @@
 ./mount_ramfs.sh ${PLATFORM}
 cd ../filesystem
 ./mount.sh 1
-sudo rm -rf fs/*
-sudo cp -rf ../rootfs/fs/* fs/
+rm -rf fs/*
+cp -rf ../rootfs/fs/* fs/
 
 # Sometimes, syncing between RAM and FS takes some time
 sleep 1
diff -Naru ../rootfs/mount_ramfs.sh ./mount_ramfs.sh
--- ../rootfs/mount_ramfs.sh	2022-12-21 19:24:46.934696399 +0100
+++ ./mount_ramfs.sh	2023-03-09 12:15:21.297374985 +0100
@@ -12,6 +12,6 @@
 # mount the rootfs
 mkdir -p fs
 
-DEVLOOP=$(sudo losetup --partscan --find --show ./board/$1/rootfs.fat)
+DEVLOOP=$(losetup --partscan --find --show ./board/$1/rootfs.fat)
 
-sudo mount ${DEVLOOP}p1 fs
+mount ${DEVLOOP}p1 fs
diff -Naru ../rootfs/umount_ramfs.sh ./umount_ramfs.sh
--- ../rootfs/umount_ramfs.sh	2022-12-21 19:24:46.934696399 +0100
+++ ./umount_ramfs.sh	2023-03-09 12:15:29.093188502 +0100
@@ -8,8 +8,8 @@
 
 echo "Here: board is $1"
  
-sudo umount fs
-sudo losetup -D
-sudo rm -rf fs
+umount fs
+losetup -D
+rm -rf fs
 
 
