--- st	2022-12-21 17:53:27.090095987 +0100
+++ ../ci/st	2022-12-21 11:07:08.951896063 +0100
@@ -1,4 +1,17 @@
-# Launch script for SO3 
+#!/bin/bash
+
+# Post-installation
+if [ ! -f post_installation_done ]; then
+
+	cd filesystem 
+	./create_img.sh vexpress && cd ..
+	cd rootfs && ./create_ramfs.sh vexpress && cd ..
+	./deploy.sh -bu
+	echo "OK" > post_installation_done
+	
+fi
+
+
 
 N_QEMU_INSTANCES=`ps -A | grep qemu | wc -l`
 GDB_PORT_BASE=1234
@@ -19,7 +32,7 @@
 
 if [ "$PLATFORM" == "virt64" ]; then
 
-sudo qemu-system-aarch64 $@ \
+qemu-system-aarch64 $@ \
   	-smp 4 -s \
 	-serial mon:stdio  \
 	-M virt,gic-version=2 -cpu cortex-a72  \
@@ -33,7 +46,8 @@
 
 else
 
-sudo qemu-system-arm $@ \
+
+qemu-system-arm $@ \
         -gdb tcp::$QEMU_GDB_PORT \
   	-smp 4 \
 	-serial mon:stdio  \
@@ -42,7 +56,6 @@
 	-kernel u-boot/u-boot \
 	-semihosting-config enable,target=native \
 	-net user,tftp=. \
-	-nic tap,script=scripts/qemu-ifup.sh,downscript=scripts/qemu-ifdown.sh,mac=${QEMU_MAC_ADDR} \
 	-sd filesystem/sdcard.img.vexpress \
 	-nographic 
 fi
