--- st	2023-02-28 20:18:22.264568940 +0100
+++ st.n	2023-03-09 12:29:19.613117491 +0100
@@ -1,5 +1,17 @@
 #!/bin/bash
 
+
+# Post-installation
+if [ ! -f post_installation_done ]; then
+
+	cd filesystem 
+	./create_img.sh vexpress && cd ..
+	cd rootfs && ./create_ramfs.sh vexpress && cd ..
+	./deploy.sh -bu
+	echo "OK" > post_installation_done
+	
+fi
+
 # Launch script for SO3 
 
 N_QEMU_INSTANCES=`ps -A | grep qemu | wc -l`
@@ -21,7 +33,7 @@
 
     if [ "$PLATFORM" == "virt64" ]; then
 
-    sudo qemu-system-aarch64 $@ ${USR_OPTION} \
+    qemu-system-aarch64 $@ ${USR_OPTION} \
  	-smp 4  \
 	-serial mon:stdio  \
 	-M virt,gic-version=2 -cpu cortex-a72  \
@@ -36,7 +48,7 @@
 
     else    
    echo Starting on ARM32
-   sudo qemu-system-arm $@ ${USR_OPTION} \
+   qemu-system-arm $@ ${USR_OPTION} \
   	-smp 4 \
 	-serial mon:stdio  \
 	-M vexpress-a15,secure=on  -cpu cortex-a15 \
