--- /home/rossierd/soo/git/infrabase/build/tmp/work/uboot-2022.04-r0/uboot-2022.04/arch/arm/mach-omap2/hwinit-common.c	2025-09-11 16:44:26.548023867 +0200
+++ ./arch/arm/mach-omap2/hwinit-common.c	2025-09-11 16:34:39.359723681 +0200
@@ -135,6 +135,16 @@
 }
 
 /**
+ * rtc32k_enable() - Detect board description
+ *
+ * Function to detect board description. This is expected to be
+ * overridden in the SoC family board file where desired.
+ */
+void __weak rtc32k_enable(void)
+{
+}
+
+/**
  * vcores_init() - Assign omap_vcores based on board
  *
  * Function to pick the vcores based on board. This is expected to be
