--- /home/rossierd/soo/git/infrabase/build/tmp/work/uboot-2022.04-r0/uboot-2022.04/include/configs/rpi.h	2025-09-11 16:44:26.946034153 +0200
+++ ./include/configs/rpi.h	2025-09-07 09:29:44.733675083 +0200
@@ -36,9 +36,12 @@
 					 CONFIG_SYS_SDRAM_SIZE - \
 					 GENERATED_GBL_DATA_SIZE)
 
+#if 0 /* heig-vd */
 #ifdef CONFIG_ARM64
 #define CONFIG_SYS_BOOTM_LEN		SZ_64M
 #endif
+#endif /* 0 */
+#define CONFIG_SYS_BOOTM_LEN		SZ_64M
 
 /* Devices */
 /* GPIO */
@@ -124,6 +127,8 @@
  * only 64M, the remaining 25M starting at 0x02700000 should allow quite
  * large initrds before they start colliding with U-Boot.
  */
+
+#if 0 /* HEIG-VD/DRE */
 #define ENV_MEM_LAYOUT_SETTINGS \
 	"fdt_high=" FDT_HIGH "\0" \
 	"initrd_high=" INITRD_HIGH "\0" \
@@ -168,12 +173,21 @@
 
 #include <config_distro_bootcmd.h>
 
+#endif /* 0 */
+
+#define CONFIG_BOOTCOMMAND \
+            "fatload mmc 0 0x02000000 uEnv.txt; env import 0x02000000; run start\0\0" \
+            "bootdelay=0\0"
+
+
+#if 0
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	"dhcpuboot=usb start; dhcp u-boot.uimg; bootm\0" \
 	ENV_DEVICE_SETTINGS \
 	ENV_DFU_SETTINGS \
 	ENV_MEM_LAYOUT_SETTINGS \
 	BOOTENV
+#endif /* 0 */
 
 
 #endif
