--- /home/rossierd/soo/git/infrabase/build/tmp/work/uboot-2022.04-r0/uboot-2022.04/include/configs/ti_omap5_common.h	2025-09-11 16:44:26.950034257 +0200
+++ ./include/configs/ti_omap5_common.h	2025-09-07 09:29:44.733637144 +0200
@@ -256,13 +256,14 @@
 		"if test $fdtfile = undefined; then " \
 			"echo WARNING: Could not determine device tree to use; fi; \0"
 
-#define BOOT_TARGET_DEVICES(func) \
-	func(MMC, mmc, 0) \
-	func(MMC, mmc, 1) \
-	func(PXE, pxe, na) \
-	func(DHCP, dhcp, na)
-
-#include <config_distro_bootcmd.h>
+#define CONFIG_BOOTCOMMAND \
+	"run findfdt; " \
+	"setenv mmcdev 0; " \
+	"setenv devtype mmc; " \
+	"run mmc_boot;" \
+	"setenv mmcdev 1; " \
+	"run mmc_boot;" \
+	""
 
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	DEFAULT_LINUX_BOOT_ENV \
@@ -270,10 +271,12 @@
 	DEFAULT_FIT_TI_ARGS \
 	DEFAULT_COMMON_BOOT_TI_ARGS \
 	DEFAULT_FDT_TI_ARGS \
+	BBAI_MMC_BOOT \
+	BBAI_UNAME_BOOT \
+	BBAI_EEPROM_PROGRAMMING \
 	DFUARGS \
 	NETARGS \
 	NANDARGS \
-	BOOTENV
 
 /*
  * SPL related defines.  The Public RAM memory map the ROM defines the
